torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -2662,4 +2692,98 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
   return work;
 }
 
+void ProcessGroupGloo::monitoredBarrier(const BarrierOptions& opts) {
+  // Use default timeout if no timeout was specified.
+  auto monitoredBarrierTimeout =
+      (opts.timeout == kUnsetTimeout) ? this->options_->timeout : opts.timeout;
+  auto rank = this->getRank();
+  auto t1 = nextTag();
+  auto t2 = nextTag();
+  std::vector<at::Tensor> commTensor = {at::tensor({rank})};
+  // only enforce timeout on rank 0. This is so that other ranks aren't timed
+  // out first, bringing down the job without reporting which rank timed out.
+  if (rank != 0) {
+    auto sendWork = send(commTensor, 0, t1);
+    auto recvWork = recv(commTensor, 0, t2);
+    sendWork->wait();
+    recvWork->wait();
+    return;
+  }
+  auto startTime = std::chrono::steady_clock::now();
+  auto worldSize = this->getSize();
+  // Holds mapping of rank to pair of recv/send work.
+  std::unordered_map<
+      int,
+      std::tuple<
+          c10::intrusive_ptr<ProcessGroup::Work>,
+          c10::intrusive_ptr<ProcessGroup::Work>>>
+      works;
+  // Kick off all work
+  for (int dstRank = 1; dstRank < worldSize; ++dstRank) {
+    auto recvWork = recv(commTensor, dstRank, t1);
+    auto sendWork = send(commTensor, dstRank, t2);
+    works.insert(
+        {dstRank, std::make_tuple(std::move(recvWork), std::move(sendWork))});
+  }
+
+  const std::string kNoRemainingTimeError = c10::str(
+      "Rank ",
+      rank,
+      " timed out in monitoredBarrier after ",
+      monitoredBarrierTimeout.count(),
+      " ms.");
+  std::vector<int> processedRanks;
+
+  // Wait for send/recv from all ranks
+  for (auto& work : works) {
```

Comment:
This is to identify first rank that did not pass monitored barrier, right?  as you mentioned, the follow up will be to find out all the ranks that did not pass monitored barriers, as we do not know who is the real first rank that did not pass monitored barrier, right?

Reply:
This is to identify a general rank that did not pass the barrier, not necessarily the first. We don't enforce an order on checking for the ranks, should we do this?

Reply:
right, I think my comment is not clear. I meant it will identify the first stuck rank in the order how works are ordered in the map or in the order when works are waited.  Did you check whether this order is always deterministic or random?

Reply:
I guess it will be random because unordered_map doesn't give any guarantee on element order. I think we should just convert it to use std::map here and then have a proper order (least to greatest rank), no. of ranks is not that big that there will be a slowdown.